diff --git a/crypto/o_init.c b/crypto/o_init.c
index a0b4256..6a32610 100644
--- a/crypto/o_init.c
+++ b/crypto/o_init.c
@@ -9,13 +9,53 @@
 
 #include "e_os.h"
 #include <openssl/err.h>
+#include <openssl/provider.h>
+
+# include <fcntl.h>
+# include <stdlib.h>
+# include <unistd.h>
+
+
+static OSSL_PROVIDER *fips, *base;
+
+static void fips_cleanup(void)
+{
+    if (base != NULL)
+        OSSL_PROVIDER_unload(base);
+    if (fips != NULL)
+        OSSL_PROVIDER_unload(fips);
+}
+
+static void fips_init(void)
+{
+    /* Not in FIPS mode; does flag exist? */
+    int fd = open("/etc/fips/enabled", O_RDONLY);
+    if (fd != -1) {
+        close(fd);
+
+        fips = OSSL_PROVIDER_load(NULL, "fips");
+        if (fips == NULL) {
+            abort();
+        }
+
+        base = OSSL_PROVIDER_load(NULL, "base");
+        if (base == NULL) {
+            OSSL_PROVIDER_unload(fips);
+            abort();
+        }
+
+        atexit(fips_cleanup);
+    }
+}
 
 /*
  * Perform any essential OpenSSL initialization operations. Currently does
  * nothing.
  */
 
-void OPENSSL_init(void)
+__attribute__((constructor)) void OPENSSL_init(void)
 {
+    fips_init();
+
     return;
 }
