Index: os-openssl/crypto/evp/pmeth_lib.c
===================================================================
--- os-openssl.orig/crypto/evp/pmeth_lib.c
+++ os-openssl/crypto/evp/pmeth_lib.c
@@ -1057,6 +1057,11 @@ static int evp_pkey_ctx_add1_octet_strin
         return 1;
     }
 
+    if (OSSL_PARAM_locate_const(EVP_PKEY_CTX_gettable_params(ctx), param) == NULL) {
+        /* Older provider that doesn't support getting this parameter */
+        return evp_pkey_ctx_set1_octet_string(ctx, fallback, param, op, ctrl, data, datalen);
+    }
+
     /* Get the original value length */
     os_params[0] = OSSL_PARAM_construct_octet_string(param, NULL, 0);
     os_params[1] = OSSL_PARAM_construct_end();
